### PR TITLE
decoder: Always map transaction.id

### LINKED ISSUE #8664: `transaction_id` in error events is read only when a transaction object is sent as well
**APM Server version** (`apm-server version`): 7.17.4, probably also in 8.x.

**Description of the problem including expected versus actual behavior**:

In [`mapToErrorModel`](https://github.com/elastic/apm-server/blob/7.17/model/modeldecoder/v2/decoder.go#L394-L405), `transaction_id` is only read when `"transaction": {}` is sent as well. 

**Steps to reproduce**:

Send this error to the events intake API with and without `"transaction":{"sampled":true,"type":"request"}`:
```
{"metadata":{"process":{"pid":1234,"title":"/usr/lib/jvm/java-10-openjdk-amd64/bin/java","ppid":1,"argv":["-v"]},"system":{"architecture":"amd64","detected_hostname":"8ec7ceb99074","configured_hostname":"host1","platform":"Linux","container":{"id":"8ec7ceb990749e79b37f6dc6cd3628633618d6ce412553a552a0fa6b69419ad4"},"kubernetes":{"namespace":"default","pod":{"uid":"b17f231da0ad128dc6c6c0b2e82f6f303d3893e3","name":"instrumented-java-service"},"node":{"name":"node-name"}}},"service":{"name":"1234_service-12a3","version":"4.3.0","node":{"configured_name":"8ec7ceb990749e79b37f6dc6cd3628633618d6ce412553a552a0fa6b69419ad4"},"environment":"production","language":{"name":"Java","version":"10.0.2"},"agent":{"version":"1.10.0","name":"java","ephemeral_id":"e71be9ac-93b0-44b9-a997-5638f6ccfc36"},"framework":{"name":"spring","version":"5.0.0"},"runtime":{"name":"Java","version":"10.0.2"}},"labels":{"group":"experimental","ab_testing":true,"segment":5}}}
{"error":{"id":"9876543210abcdeffedcba0123456789","timestamp":1571657444929001,"trace_id":"0123456789abcdeffedcba0123456789","parent_id":"9632587410abcdef","transaction_id":"1234567890987654","transaction":{"sampled":true,"type":"request"},"culprit":"opbeans.controllers.DTInterceptor.preHandle(DTInterceptor.java:73)","log":{"message":"Request method 'POST' not supported","param_message":"Request method 'POST' /events/:event not supported","logger_name":"http404","level":"error","stacktrace":[{"abs_path":"/tmp/Socket.java","filename":"Socket.java","classname":"Request::Socket","function":"connect","vars":{"key":"value"},"pre_context":["line1","line2"],"context_line":"line3","library_frame":true,"lineno":3,"module":"java.net","colno":4,"post_context":["line4","line5"]},{"filename":"SimpleBufferingClientHttpRequest.java","lineno":102,"function":"executeInternal","abs_path":"/tmp/SimpleBufferingClientHttpRequest.java","vars":{"key":"value"}}]},"exception":{"message":"Theusernamerootisunknown","type":"java.net.UnknownHostException","handled":true,"module":"org.springframework.http.client","code":42,"handled":false,"attributes":{"foo":"bar"},"cause":[{"type":"InternalDbError","message":"something wrong writing a file","cause":[{"type":"VeryInternalDbError","message":"disk spinning way too fast"},{"type":"ConnectionError","message":"on top of it,internet doesn't work"}]}],"stacktrace":[{"abs_path":"/tmp/AbstractPlainSocketImpl.java","filename":"AbstractPlainSocketImpl.java","function":"connect","vars":{"key":"value"},"pre_context":["line1","line2"],"context_line":"3","library_frame":true,"lineno":3,"module":"java.net","colno":4,"post_context":["line4","line5"]},{"filename":"AbstractClientHttpRequest.java","lineno":102,"function":"execute","vars":{"key":"value"}}]},"context":{"request":{"socket":{"remote_address":"12.53.12.1","encrypted":true},"http_version":"1.1","method":"POST","url":{"protocol":"https:","full":"https://www.example.com/p/a/t/h?query=string#hash","hostname":"www.example.com","port":8080,"pathname":"/p/a/t/h","search":"?query=string","hash":"#hash","raw":"/p/a/t/h?query=string#hash"},"headers":{"Forwarded": "for=192.168.0.1", "host":"opbeans-java:3000","content-length":"0","cookie":["c1=v1","c2=v2"],"Elastic-Apm-Traceparent":"00-8c21b4b556467a0b17ae5da959b5f388-31301f1fb2998121-01"},"cookies":{"c1":"v1","c2":"v2"},"env":{"SERVER_SOFTWARE":"nginx","GATEWAY_INTERFACE":"CGI/1.1"},"body":"HelloWorld"},"response":{"status_code":200,"headers":{"content-type":"application/json"},"headers_sent":true,"finished":true},"user":{"id":99,"username":"foo","email":"user@foo.mail"},"tags":{"organization_uuid":"9f0e9d64-c185-4d21-a6f4-4673ed561ec8"},"custom":{"my_key":1,"some_other_value":"foobar","and_objects":{"foo":["bar","baz"]}},"service":{"name":"service1","node":{"configured_name":"node-xyz"},"language":{"version":"1.2"},"framework":{"version":"1","name":"Node"}}}}}
```
One of the created document won't have a `transaction.id` field:
```
GET apm-*-error-*/_search?filter_path=**.id
{
  "query": {
    "term": {
      "container.id": "8ec7ceb990749e79b37f6dc6cd3628633618d6ce412553a552a0fa6b69419ad4"
    }
  }
}
```
```
{
  "hits" : {
    "hits" : [
      {
        "_source" : {
          "container" : {
            "id" : "8ec7ceb990749e79b37f6dc6cd3628633618d6ce412553a552a0fa6b69419ad4"
          },
          "parent" : {
            "id" : "9632587410abcdef"
          },
          "error" : {
            "id" : "9876543210abcdeffedcba0123456789"
          },
          "observer" : {
            "id" : "33b881e1-1889-4bf6-be7d-00de3c207eee"
          },
          "trace" : {
            "id" : "0123456789abcdeffedcba0123456789"
          },
          "user" : {
            "id" : "99"
          },
          "transaction" : {
            "id" : "1234567890987654"
          }
        }
      },
      {
        "_source" : {
          "container" : {
            "id" : "8ec7ceb990749e79b37f6dc6cd3628633618d6ce412553a552a0fa6b69419ad4"
          },
          "parent" : {
            "id" : "9632587410abcdef"
          },
          "error" : {
            "id" : "9876543210abcdeffedcba0123456789"
          },
          "observer" : {
            "id" : "33b881e1-1889-4bf6-be7d-00de3c207eee"
          },
          "trace" : {
            "id" : "0123456789abcdeffedcba0123456789"
          },
          "user" : {
            "id" : "99"
          }
        }
      }
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by marclop**
## Motivation/summary

Removes the need for the `transaction` object to be set in the intake
payload in order to map the `transaction_id` to `transaction.id`. Any
incoming events with `transaction_id` set, but `transaction` unset will
now correctly have `transaction.id` set.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
~- [ ] Documentation has been updated~

## How to test these changes

1. `docker-compose up -d`.
2. `cd systemtest && go run cmd/runapm/main.go`
3. `curl -H 'Content-type: application/x-ndjson' --data-binary @testdata/intake-v2/errors_transaction_id.ndjson "http://localhost:60159/intake/v2/events"` Replace port `60159` with the actual port from `runapm`.
4. Issue the following query and ensure it matches the document:
```
GET /logs-apm.error-default/_search
{
  "fields": [
    "transaction.id"
  ], 
  "_source": false, 
  "query": {
    "match": {
      "transaction.id": "1234567890987654"
    }
  }
}
```

```javascript
{
  "took": 9,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 0.2876821,
    "hits": [
      {
        "_index": ".ds-logs-apm.error-default-2022.08.08-000001",
        "_id": "AAKlfIIByzLEhspeH8SC",
        "_score": 0.2876821,
        "fields": {
          "transaction.id": [
            "1234567890987654"
          ]
        }
      }
    ]
  }
}
```

## Related issues

Closes #8664

### CODE REVIEW

File: changelogs/8.4.asciidoc [hunk 1]
```diff
@@ -15,6 +15,7 @@ https://github.com/elastic/apm-server/compare/8.3\...8.4[View commits]
 ==== Bug fixes
 - Fix race for deducing destination service fields for OTel bridge {pull}8363[8363]
 - Fix processors receiving events even after the server has stopped {pull}8388[8388]
+- Fix a bug where an event's transaction_id is ignored if no transaction object is set {pull}XXXX[XXXX]
```
Comment (marclop) [inline changelogs/8.4.asciidoc:18]:
```suggestion
- Fix a bug where an event's transaction_id is ignored if no transaction object is set {pull} 8820[8820]
```

**Comment by marclop** [comment]
Added backport `8.4`, but we're way past the FF date, so we may not want to merge the backport PR for that branch until we ship `8.4.0`.

**Review by axw — APPROVED** [review]

**Comment by axw** [comment]
@marclop I think we should include this in 8.4.0.

**Review by lahsivjar — APPROVED** [review]
LGTM, I think we can merge for 8.4.0 too

**Comment by axw** [comment]
Confirmed with 8.5.0-BC1. I created an Elastic Cloud deployment, removed the APM Server version check (because the integration package hasn't been updated to 8.5.0 yet),  sent `errors_transaction_id.ndjson`, and verified the error event has `transaction.id` set correctly.

**Pull request merged**